### PR TITLE
Fix ro segments with regions

### PR DESCRIPTION
**Pull request by PeterSolMS**
There were a couple places in the mark and relocation phases where the regions logic couldn't cope with objects outside of the memory range reserved for the GC heap. These objects may occur in read-only segments (but only if FEATURE_BASICFREEZE is enabled).

Perhaps we should centralize the test for a valid object pointer in a (inlineable) helper though?

### CODE REVIEW

**Review by mangod9 — APPROVED** [review]

**Review by Maoni0 — CHANGES_REQUESTED** [review]
please do include the `#ifdef FEATURE_BASICFREEZE` check as this is very intentional - if we don't have this feature defined we wouldn't need to do these checks at all (and this feature isn't necessarily defined (or will not be defined) everywhere).

also nit - more ()'s :)

File: src/coreclr/gc/gc.cpp [hunk 1]
```diff
@@ -22286,13 +22286,29 @@ BOOL gc_heap::gc_mark1 (uint8_t* o)
     return marked;
 }
 
+#ifdef USE_REGIONS
+inline bool in_heap_range (uint8_t* o)
+{
+#ifdef FEATURE_BASICFREEZE
+    // we may have frozen objects in read only segments
+    // outside of the reserved address range of the gc heap
+    return ((g_gc_lowest_address <= o) && (o < g_gc_highest_address));
```
Comment (cshung) [inline src/coreclr/gc/gc.cpp:22295]: Should we also check, in debug mode, that the pointer lies in a read-only segment? That assertion could be helpful to find real bad pointers.

**Review by cshung — APPROVED** [review]
It looks like none of our many tests caught this issue, it would be nice if we could add a test will.

File: src/coreclr/gc/gc.cpp (hunk 1 above)
Comment (PeterSolMS) [reply src/coreclr/gc/gc.cpp:22295]: Good point - I will add an assert before the return statement:

    assert (((g_gc_lowest_address <= o) && (o < g_gc_highest_address)) || (ro_segment_lookup(o) != nullptr));

**Comment by PeterSolMS** [comment]
I'll try to find out how we can construct a test case for read-only segments...

**Review by Maoni0 — APPROVED** [review]
LGTM!

**Pull request merged**